### DISCUSSION TITLE
Certificate signing flow for embedded/iot devices

### DISCUSSION THREAD

**Discussion by daFritz84** [smallstep/certificates#812] [Q&A]
Hi,

so I'm trying to craft a viable certificate signing flow using smallstep for our embedded device portfolio. The devices themselve are quite beefy, think raspberry pi equivalent HW. One exceptional feature is the presence of a HSM/TPM module.

So I want to use the JWK provisioner this way:
1. Smallstep CA is set up on the server/cloud
2. Root certificate is distributed via signed FW image (we have secure/trusted boot enabled)
3. JWK key pair is generated inside the HSM
4. JWK public key is uploaded to the CA and a new provisioner is created for the device. This is kind of important, since the uploading user basically vouches for the device. Think of this like a device onboarding flow, where mutual trust is established by the user uploading the public key and deploying the root certificate.
5. the device can generate its csr, private key is again created inside the HSM
6. finally create token & get the cert signed
7. live happiliy ever after, rinse and repeat step 5-6 to renew/recreate

What I want to know if this is feasible after all. I'm especially worried about the number of provisioners performancewise, i.e. 10k+. But I don't want an encrypted JWK key, where I need to distribute the pw to every device.

Any suggestion to improve this approach are highly welcome. Also if anyone has ideas for a different approach, please advise.

Thanks for any feedback,
Stefan

**Comment by tashian** [comment]
Hi! This is an interesting flow that I haven't seen before.

I think the big question I have is around step 4 and how that key exchange is going to be authenticated. It would need to happen outside of the CA, but thankfully you can add the provisioner without restarting `step-ca`, either by using the new database-backed provisioner management system, or by adding it to ca.json and sending a `HUP` to step-ca. And, you may see different performance characteristics (in terms of # of provisioners) between the two approaches. I'm not sure if we've tested 10k provisioners yet, but one immediate concern I'd have is that the `/provisioners` endpoint is not paginated, so with 10k provisioners you'd get a ~10MB JSON response from that endpoint.

An alternative to JWKs here is to use the X5C provisioner.
The idea with X5C is that the device itself gets a birth certificate baked into the firmware, along with your main PKI's root cert.
The birth certificate lives under a different PKI / root, and the key for it could be generated in your TPM.
That PKI's root is configured in the X5C provisioner.
When the device comes up, it requests a cert using its birth certificate.

Do you know if your TPM comes with a built-in device certificate? Some do, some don't, but it may be something you could leverage. That certificate could serve as the birth certificate for the device, and you'd configure its root as the X5C provisioner root. The big challenge with this approach (using the TPM's built-in cert) is that you'd also need to maintain a list of device IDs that you can filter by on the CA side. That would be a bit complicated to do with `step-ca` templates, if it's possible at all (@maraino will know more here). We have only just added the ability to access the X5C authorization certificate contents in templates—it's still a pre-release feature as part of our [nebula provisioner support](https://github.com/smallstep/certificates/pull/785).

What you don't get in either of these cases is any sort of device attestation, and I wonder if it's possible to incorporate that.
But if you're using secure boot, that's a good step in terms of integrity.

I'm curious what options other folks see for this.

**Reply by maraino** [reply]
We need a few changes to have access to the certificate in the templates using the X5C provisioner. If you have an attestation root, you would be able to use the X5C provisioner. Right now we don't have support for inventories to only allow certificates of a fixed set of devices. Something can be done in the templates, but it can be hard to manage unless, for example, the attestation certificate has some SAN that we can validate (i.e. company identifier). 

Regarding JWK, I don't know step-ca will perform with 10k+ provisioners. The JWK private key is not required by step-ca, the encrypted version is to make the distribution simpler. If you can distribute one or a set of keys to your HSMs/TPMs, you can use those to sign a token, instead of generating one for each device.

Another idea, that comes to mind is to use some kind of OIDC IdP for device inventory, and try somehow to authenticate those devices with the IdP to get an OIDC token. For SSH certificates we require the `"email"` claim, but that's not the case for X.509 ones. For X.509, the default SANs on an OIDC provisioner are the email address (if available) and the issuer URI, but as with templates you have access to the CSR you can use the SANs in the CSR to generate the desired certificate.

**Comment by daFritz84** [comment]
Thanks for the replies so far. I guess, the JWK method will not work as hoped.

As for the X5C, while feasible (the HSM supports certificates), we want to avoid a common root/attestation CA for device certificates at all costs. So the underlying problem remains the same, I would need some kind of inventory (like @maraino mentioned). Apart from the fact, that I don't want a common attestion root for the devices, I can add any SAN to a device attestion certificate. I'm not familiar with the templates yet, basically I just started learning, but I suppose this does not help the X5C case?

The OIDC IdP is quite an interesting proposal since it would allow for a clean cut between the CA and Authorization/Authentication functionality. However, it does add a lot of additional complexity. I'm also not quite sure if some OIDC provider off the shelf would work. So I would be running a custom OIDC provider which sounds like a big no no to me.

In the end I would be happy to write my own inventory with some verification endpoint for jwt tokens. I'm thinking like a modified version of the JWK approch, but instead of working with a single key, I would provide some lookup table for different public keys. I.e. single provisioner with a multiple key map. Does this make sense at all?

**Reply by tashian** [reply]
Yeah this makes sense.

If you don't want to go full OIDC, you could write a token server that knows how to authenticate your users and that signs JWTs that the CA's `/sign` endpoint will accept (along with a matching CSR). The token server holds the JWK private key, and you would delegate user authentication / authorization to it however you like.

Two examples of this in practice:

- The [smallstep/clients repo](https://github.com/smallstep/clients) has some example code showing how to construct a request for the `/sign` endpoint.
-  The `step ca token` command will generate a one-time-use token that the CA `/sign` endpoint will accept (you can pass it into `step ca sign` later). These tokens have a 5 minute validity period.

**Comment by daFritz84** [comment]
Addendum: Just a little bit of background information, why I'm opposed to a root attestation certificate:
So basically, we are in the business of industrial control systems were we have strong ties to the person who integrates/programs our controllers. The user uploading the attestation cert/jwk pub key (in the sense of my proposed scheme) to the cloud is in fact the integrator and is the most trusted person in the manufacturing/deployment chain (the bare device is useless without the program). So I don't want to onboard any device to the cloud, but only devices where a user personally has uploaded some kind of token and I can associate this device to the corresponding account. Anyways, I appreciate the feedback and suggestions.

**Reply by tashian** [reply]
Ahh I see. It sounds like OIDC could be a good option here because you're authenticating a person/account, and then that person is effectively attesting the device by programming it.

**Reply by mmalone** [reply]
We've worked with other users/customers that have similar requirements. I think the right provisioner for this scenario might be OIDC, not JWK or X5C. The trusted user authenticates to an identity provider and gets an OIDC identity token. The certificate signing request (CSR) is submitted to the CA along with the identity token, and the identity token is used to authenticate the user and authorize the certificate request.

This flow can be implemented using `step ca certificate` and selecting an OIDC provisioner. You can also get the OIDC token in a separate step (e.g., using `step ca token`, or some other mechanism) and pass it in to `step ca certificate` via the `--token` parameter.

The big limitation of this approach is that the only information the CA has to authenticate & authorize the request is the OIDC identity token. That leaves you with two options:

1. The certificate details (e.g., subject name) can be derived from information available in the OIDC identity token (this can be customized using templates).
2. The CA can allow specific users to specify any certificate details they want (this can be implemented by marking users as "admins" on the OIDC provisioner).

If you _really_ trust your integrators, the latter may be sufficient. Alternatively, you could use custom claims in an OIDC identity token to control certificate details (i.e., you can put whatever you want in your OIDC identity tokens). Any standards-compliant off-the-shelf OAuth OIDC implementation should work here. Frankly, the CA doesn't require much... you just need an [OIDC configuration endpoint](https://accounts.google.com/.well-known/openid-configuration) that contains a [`jwks_uri`](https://www.googleapis.com/oauth2/v3/certs) with the public keys that the CA should use to validate identity tokens.

There are other options... you could also build a JWK-based "registration authority" that sits between your clients & the CA, as @tashian suggested above. Ultimately, though, it sounds like you're gonna need some system that authenticates the user (and/or device) out-of-band and issues a credential that `step-ca` understands with attributes that map into certificate details (using templates).

We are working on several features that I think are relevant here and could make this easier/better. I'm not sure how much of this is gonna land in open source vs. our commercial product, but here's the list:

- **Lifecycle webhooks** will let you configure `step-ca` to call an external web service to provide metadata that can be referenced in certificate templates to customize cert details, authorize certificate requests, and for audit logging.
- **Inventories** will allow you to build entity databases that can be used to customize certificate details (e.g., mapping a device ID to an account, region, or device name).
- **TPM attestations** a new provisioner that makes it easier to leverage standards-based TPM attestations to authenticate certificate requests.
- **Factor sequencing** will allow you to require multiple credentials to authenticate a certificate request (e.g., an OIDC identity token + a TPM attestation).

If you're interested in discussing any of these features lemme know.

**Comment by daFritz84** [comment]
@tashian absolutely correct, attestation happens by programming. Without programming our devices are pretty useless 😄 

Anyways, we do have an IdP for our users in place (keycloak), but if possible, I would like to keep the devices as separate entities (associated with project entities, which in turn are associated with user entities).

@mmalone: I would go for my own `registration authority` interfacing with `step-ca`. But no matter which provisioner I choose (`oidc`, `jwk`, `x5c`, ...) I'm ending up banging my head on the fact that `step-ca` needs a single public key. There are, of course, ways around this problem. I.e. I could run a `registration authority` which issues token derived from a single key to each device upon authenticating with some out-of-band exchanged secrets (e.g. another set of JWKs). It just seems that I need to add a lot of unnecessary complexity, and in the end it would be a lot easier if there was some officially supported plugin interface for provisioners. Any thoughts on this?

**Reply by maraino** [reply]
> it would be a lot easier if there was some officially supported plugin interface for provisioners. Any thoughts on this?

That's an idea I mentioned some time ago, but we haven't implemented it yet. We didn't have a use case at the time. What kind of provisioner are you thinking of?

**Reply by daFritz84** [reply]
I haven't had time to do a POC to test the feasability, but I would do a variation of the JWK provisioner for my use case like that:
Basically, instead of having a single key I would allow for a (large) set of keys. I could add some SAN or other metadata to the CSR or token generation to allow for a key lookup. Keys would be stored in a db backend. I'm not sure how I would do key management yet.

So I would do the following to sign a certificate:
1. Generate the jwk key pair on the device.
2. Leave the private key in the HSM, transfer the public key including a identifiying SAN (e.g.uri) out-of-band to the cloud. 
3. public key is added to the provisioners keystore/db
4. Generate Token on the device and csr with SAN-id
5. Sign, proper pub-key to validate the token is found via SAN-id 

I reckon that the described provisioner solution is not something which would be accepted  in upstream `step-ca`. In the end, its a niche solution for a niche problem. But a plugin interface would allow me to maintain my own solution separate from the official `step-ca`.

**Reply by maraino** [reply]
Totally, a way to plug provisioners, even if it is just re-compiling step-ca would be super useful for your use case. Something like `step-ca --plugin plugin.so` has some extra problems, because it won't work on some platforms, but might be nice to support that too :)
